File: py_common/base_convert.py
#!/usr/bin/env python3
"""
Base-10 and Base-32 Converter

Command-line tool for converting between base-10 and base-32 (MPFR format).
Uses the mpfr_base32 module with gmpy2 for MPFR-compatible operations.
"""

import sys
import argparse

from mpfr_base32 import parse_mpfr_base32, decimal_to_mpfr_base32  # type: ignore


def remove_trailing_zeros(s: str) -> str:
    """
    Remove trailing zeros from a number string that has a decimal point.
    Also removes the decimal point if no fractional part remains.
    
    Args:
        s: Number string
    
    Returns:
        String with trailing zeros removed
    """
    if '.' not in s:
        return s
    
    # Remove trailing zeros
    s = s.rstrip('0')
    # Remove decimal point if it's now at the end
    s = s.rstrip('.')
    return s


def convert_10_to_32(base10_str: str, precision_bits: int = 256) -> str:
    """
    Convert base-10 string to base-32 (MPFR format).
    
    Args:
        base10_str: Number in base-10 format (e.g., "-0.5", "123.456", "1e-10")
        precision_bits: Precision in bits (like C's mpfr_prec_t)
    
    Returns:
        Number in base-32 MPFR format
    """
    try:
        # Convert to base-32 using gmpy2
        result = decimal_to_mpfr_base32(base10_str, precision_bits)
        return result
    except Exception as e:
        raise ValueError(f"Invalid base-10 number: {e}")


def convert_32_to_10(base32_str: str, precision_bits: int = 256) -> str:
    """
    Convert base-32 (MPFR format) string to base-10.
    
    Args:
        base32_str: Number in base-32 format (e.g., "-0.g", "a")
        precision_bits: Precision in bits (like C's mpfr_prec_t)
    
    Returns:
        Number in base-10 format with trailing zeros
    """
    import math
    
    try:
        # Parse base-32 string to mpfr with specified precision
        value = parse_mpfr_base32(base32_str, precision_bits)
        
        # Handle zero
        if value == 0:
            return "0"
        
        # Convert bits to total decimal digits using the formula:
        # digits = ceil(bits * log(2) / log(10)) + 1
        # This matches MPFR's digit generation behavior
        total_digits = math.ceil(precision_bits * math.log(2) / math.log(10)) + 1
        
        # Use gmpy2's digits() function to get base-10 representation
        # This directly calls MPFR's mpfr_get_str, matching C behavior exactly
        # digits() returns (mantissa_str, exponent, precision_used)
        mantissa_str, exp, _ = value.digits(10, 0)  # 0 means use all available precision
        
        # Handle special cases
        if mantissa_str in ['@NaN@', '@Inf@', '-@Inf@']:
            return mantissa_str
        
        # mantissa_str is the mantissa without decimal point
        # exp is the exponent (number of digits before decimal point)
        sign = '-' if mantissa_str.startswith('-') else ''
        if sign:
            mantissa_str = mantissa_str[1:]  # Remove sign from mantissa
        
        mantissa_len = len(mantissa_str)
        
        # Calculate how many digits we need to match C's output
        # Pad or truncate to match the expected precision
        if mantissa_len < total_digits:
            mantissa_str = mantissa_str + '0' * (total_digits - mantissa_len)
        elif mantissa_len > total_digits:
            mantissa_str = mantissa_str[:total_digits]
        
        # Format the number with decimal point
        if exp > 0:
            # Positive exponent: digits before decimal point
            if exp >= len(mantissa_str):
                # Integer with trailing zeros
                result = mantissa_str + '0' * (exp - len(mantissa_str))
            else:
                # Insert decimal point
                integer_part = mantissa_str[:exp]
                fractional_part = mantissa_str[exp:]
                result = integer_part + '.' + fractional_part
        elif exp == 0:
            # All digits after decimal point
            result = '0.' + mantissa_str
        else:
            # Negative exponent: leading zeros after decimal point
            result = '0.' + '0' * (-exp) + mantissa_str
        
        result = sign + result
        # Remove trailing zeros before returning
        return remove_trailing_zeros(result)
    except Exception as e:
        raise ValueError(f"Invalid base-32 number: {e}")


def main():
    parser = argparse.ArgumentParser(
        description='Base-10 and Base-32 Converter - Convert between base-10 and MPFR base-32 format',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Commands:
  10TO32  Convert base-10 to base-32
  32TO10  Convert base-32 to base-10

Examples:
  %(prog)s 10TO32 64 -0.5
  %(prog)s 32TO10 64 -0.g
  %(prog)s 10TO32 128 0.25
  %(prog)s 32TO10 128 0.8
  %(prog)s 10TO32 128 123.456
  %(prog)s 32TO10 128 3r.efdf8
  %(prog)s 10TO32 256 1e-10
  %(prog)s 32TO10 256 0.00001
        """
    )
    
    parser.add_argument('command', 
                        choices=['10TO32', '32TO10'],
                        help='Conversion command')
    parser.add_argument('precision',
                        type=int,
                        help='Precision in bits for 10TO32, or decimal places for 32TO10')
    parser.add_argument('number',
                        help='Number to convert')
    
    args = parser.parse_args()
    
    try:
        if args.command == '10TO32':
            # For 10TO32, precision is in bits (like C's mpfr_prec_t)
            result = convert_10_to_32(args.number, args.precision)
            print(result)
        elif args.command == '32TO10':
            result = convert_32_to_10(args.number, args.precision)
            print(result)
        else:
            print(f"ERROR: Unknown command '{args.command}'", file=sys.stderr)
            parser.print_help()
            sys.exit(1)
    except ValueError as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"ERROR: Conversion failed: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: c_cal/base_convert.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <mpfr.h>
#include <ctype.h>
#include "mpfr_base32.h"

#define MAX_LINE_LENGTH 4096
#define DEFAULT_PRECISION 256

/**
 * Remove trailing zeros from a number string that has a decimal point
 * Also removes the decimal point if no fractional part remains
 */
void remove_trailing_zeros(char *str) {
    // Find decimal point
    char *decimal = strchr(str, '.');
    if (decimal == NULL) {
        return;  // No decimal point, nothing to do
    }
    
    // Find end of string
    char *end = str + strlen(str) - 1;
    
    // Remove trailing zeros
    while (end > decimal && *end == '0') {
        *end = '\0';
        end--;
    }
    
    // Remove decimal point if it's now at the end
    if (*end == '.') {
        // Check if the character before the decimal point is a digit
        if (end > str && isdigit(*(end - 1))) {
            *end = '\0';
        }
    }
}

/**
 * Convert base-10 to base-32
 */
void convert_10_to_32(const char *base10_str, mpfr_prec_t precision) {
    mpfr_t value;
    mpfr_init2(value, precision);
    
    // Parse base-10 string
    int result = mpfr_set_str(value, base10_str, 10, MPFR_RNDN);
    
    if (result != 0) {
        printf("ERROR: Invalid base-10 number\n");
        mpfr_clear(value);
        return;
    }
    
    // Convert to base-32
    char *base32_str = mpfr_to_base32(value);
    
    if (base32_str == NULL) {
        printf("ERROR: Conversion failed\n");
        mpfr_clear(value);
        return;
    }
    
    printf("%s\n", base32_str);
    
    free(base32_str);
    mpfr_clear(value);
}

/**
 * Convert base-32 to base-10
 */
void convert_32_to_10(const char *base32_str, mpfr_prec_t precision) {
    mpfr_t value;
    mpfr_init2(value, precision);
    
    // Parse base-32 string
    int result = parse_base32_to_mpfr(base32_str, value, precision);
    
    if (result != 0) {
        printf("ERROR: Invalid base-32 number\n");
        mpfr_clear(value);
        return;
    }
    
    // Convert to base-10 string
    mpfr_exp_t exp;
    char *base10_str = mpfr_get_str(NULL, &exp, 10, 0, value, MPFR_RNDN);
    
    if (base10_str == NULL) {
        printf("ERROR: Conversion failed\n");
        mpfr_clear(value);
        return;
    }
    
    // Handle special cases
    if (strcmp(base10_str, "@NaN@") == 0 || strcmp(base10_str, "@Inf@") == 0 || strcmp(base10_str, "-@Inf@") == 0) {
        printf("%s\n", base10_str);
        mpfr_free_str(base10_str);
        mpfr_clear(value);
        return;
    }
    
    // Handle zero
    if (mpfr_zero_p(value)) {
        printf("0\n");
        mpfr_free_str(base10_str);
        mpfr_clear(value);
        return;
    }
    
    // Format the output with decimal point
    size_t len = strlen(base10_str);
    int sign_offset = (base10_str[0] == '-') ? 1 : 0;
    size_t mantissa_len = len - sign_offset;
    
    // Print sign if present
    if (sign_offset) {
        printf("-");
    }
    
    // Build output string in a buffer for trailing zero removal
    char output[4096];
    char *out_ptr = output;
    
    if (exp > 0) {
        if ((size_t)exp >= mantissa_len) {
            // Integer with trailing zeros
            out_ptr += sprintf(out_ptr, "%s", base10_str + sign_offset);
            for (long i = mantissa_len; i < exp; i++) {
                *out_ptr++ = '0';
            }
            *out_ptr = '\0';
        } else {
            // Insert decimal point
            for (long i = 0; i < exp; i++) {
                *out_ptr++ = base10_str[sign_offset + i];
            }
            *out_ptr++ = '.';
            strcpy(out_ptr, base10_str + sign_offset + exp);
        }
    } else {
        // Leading zeros
        out_ptr += sprintf(out_ptr, "0.");
        for (long i = 0; i < -exp; i++) {
            *out_ptr++ = '0';
        }
        strcpy(out_ptr, base10_str + sign_offset);
    }
    
    // Remove trailing zeros before printing
    remove_trailing_zeros(output);
    printf("%s\n", output);
    
    mpfr_free_str(base10_str);
    mpfr_clear(value);
}

/**
 * Print usage information
 */
void print_usage(const char *program_name) {
    printf("Usage: %s <command> [options]\n\n", program_name);
    printf("Commands:\n");
    printf("  10TO32 <precision> <base10_number>  Convert base-10 to base-32\n");
    printf("  32TO10 <precision> <base32_number>  Convert base-32 to base-10\n\n");
    printf("Options:\n");
    printf("  <precision>     Precision in bits (e.g., 64, 128, 256)\n");
    printf("  <base10_number> Number in base-10 format (e.g., -0.5, 123.456, 1e-10)\n");
    printf("  <base32_number> Number in base-32 format (e.g., -0.g, a, 0.8@-1)\n\n");
    printf("Examples:\n");
    printf("  %s 10TO32 64 -0.5\n", program_name);
    printf("  %s 32TO10 64 -0.g\n", program_name);
    printf("  %s 10TO32 128 0.25\n", program_name);
    printf("  %s 32TO10 128 0.8\n", program_name);
}

/**
 * Main function
 */
int main(int argc, char *argv[]) {
    if (argc < 4) {
        print_usage(argv[0]);
        return 1;
    }
    
    const char *command = argv[1];
    long precision = atol(argv[2]);
    const char *number = argv[3];
    
    if (precision <= 0) {
        printf("ERROR: Invalid precision\n");
        return 1;
    }
    
    if (strcmp(command, "10TO32") == 0) {
        convert_10_to_32(number, precision);
    } else if (strcmp(command, "32TO10") == 0) {
        convert_32_to_10(number, precision);
    } else {
        printf("ERROR: Unknown command '%s'\n", command);
        print_usage(argv[0]);
        return 1;
    }
    
    return 0;
}
